### PR TITLE
[RISCV][GlobalISel] Lower calls to variadic functions

### PR DESCRIPTION
**Pull request by nitinjohnraj**
Calls to variadic functions do not seem to require any special handling.

### CODE REVIEW

**Comment by michaelmaitland** [comment]
In ISelLowering it looks like we do give some care to varargs.

* If all registers are allocated, then all varargs must be passed on the stack and don't need to save any argregs
* record the frame index of the first variable argument which is a value necessary to VASTART.
* if saving an odd number of registers then create an extra stack slot to ensure that the frame pointer is correctly aligned
* copy the integer register that may have been used for passing varargs to the vararg save area

Why don't we need this in GISel?

File: llvm/test/CodeGen/RISCV/GlobalISel/call-lowering/variadic-call.ll [hunk 1]
```diff
@@ -0,0 +1,15 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+; RUN: llc -mtriple=riscv32 -global-isel --stop-before=legalizer < %s -o - \
+; RUN:   | FileCheck %s
+; RUN: llc -mtriple=riscv64 -global-isel --stop-before=legalizer < %s -o - \
+; RUN:   | FileCheck %s
+
```
Comment (michaelmaitland) [inline llvm/test/CodeGen/RISCV/GlobalISel/call-lowering/variadic-call.ll:6]: need `noundef` and `signext` here and below?

File: llvm/test/CodeGen/RISCV/GlobalISel/call-lowering/variadic-call.ll [hunk 2]
```diff
@@ -0,0 +1,15 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+; RUN: llc -mtriple=riscv32 -global-isel --stop-before=legalizer < %s -o - \
+; RUN:   | FileCheck %s
+; RUN: llc -mtriple=riscv64 -global-isel --stop-before=legalizer < %s -o - \
+; RUN:   | FileCheck %s
```
Comment (michaelmaitland) [inline llvm/test/CodeGen/RISCV/GlobalISel/call-lowering/variadic-call.ll:5]: It looks like CHECK statements are missing?

File: llvm/test/CodeGen/RISCV/GlobalISel/call-lowering/variadic-call.ll [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+declare noundef signext i32 @foo(i32 noundef signext, ...)
+
+define dso_local noundef signext i32 @main() {
```
Comment (michaelmaitland) [inline llvm/test/CodeGen/RISCV/GlobalISel/call-lowering/variadic-call.ll:9]: need `dso_local`?

**Comment by topperc** [comment]
> In ISelLowering it looks like we do give some care to varargs.
> 
> * If all registers are allocated, then all varargs must be passed on the stack and don't need to save any argregs
> * record the frame index of the first variable argument which is a value necessary to VASTART.
> * if saving an odd number of registers then create an extra stack slot to ensure that the frame pointer is correctly aligned
> * copy the integer register that may have been used for passing varargs to the vararg save area
> 
> Why don't we need this in GISel?

I believe the code you mention is in RISCVTargetLowering::LowerFormalArguments in SelectionDAG. That is used to codegen the beginning of a variadic function. That will need to be done in `RISCVCallLowering::lowerFormalArguments` for gisel.

This patch is for calls to variadic functions. The equivalent is RISCVTargetLowering::LowerCall in SelectionDAG.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**